### PR TITLE
Concat Threshold Update 

### PR DESCRIPTION
**Pull request by samuel-wj-chapman**
## Pull Request Description:
feature to force the alignment of thresholds pre and post concatenation layer. implemented for both torch and keras. Including testing for each.

update does not occur when previous layer has multiple outputs.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# Copyright 2022 Sony Semiconductor Israel, Inc. All rights reserved.
+#
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py:2]: Please fix the license year.

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+from tensorflow.keras.layers import Concatenate
+
+from model_compression_toolkit.core import common
+from model_compression_toolkit.core.common import Graph, BaseNode
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.constants import THRESHOLD
+
+MATCHER = NodeOperationMatcher(Concatenate)
+
+class threshold_updater(common.BaseSubstitution):
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py:26]: Please rename the class to include 'concat' or something like that (for example: ConcatThresholdUpdate)

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+
+from tensorflow.keras.layers import Concatenate
+
+from model_compression_toolkit.core import common
+from model_compression_toolkit.core.common import Graph, BaseNode
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.constants import THRESHOLD
+
+MATCHER = NodeOperationMatcher(Concatenate)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py:24]: What about tf.concat?

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            Graph after applying the substitution.
+        """
+
+        prev_nodes = graph.get_prev_nodes(node)
+        concat_threshold = node.candidates_quantization_cfg[0].activation_quantization_cfg.activation_quantization_params.get(THRESHOLD)
+
+        for prev_node in prev_nodes:
+            if len(graph.get_next_nodes(prev_node))==1:
+                prev_node.candidates_quantization_cfg[0].activation_quantization_cfg.activation_quantization_params[THRESHOLD] = concat_threshold
+
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py:58]: We also need to deal with mixed-precision case since the threshold can be different for each cfg in candidates_quantization_cfg.

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/concat_threshold_update.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2022 Sony Semiconductor Israel, Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/concat_threshold_update.py:3]: All comments in the keras implementation are relevant here as well.

File: model_compression_toolkit/core/common/quantization/quantization_config.py [hunk 6]
```diff
@@ -62,7 +62,8 @@ def __init__(self,
                  residual_collapsing: bool = True,
                  shift_negative_ratio: float = 0.05,
                  shift_negative_threshold_recalculation: bool = False,
-                 shift_negative_params_search: bool = False):
+                 shift_negative_params_search: bool = False,
+                 concat_threshold_update: bool = False):
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/quantization/quantization_config.py:66]: Missing doc for the new argument.

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 7]
```diff
@@ -571,13 +573,21 @@ def test_qat(self):
...
         This test checks the quantization of BatchNorm layer attributes.
         """
         BNAttributesQuantization(self, quantize_linear=False).run_test()
         BNAttributesQuantization(self, quantize_linear=True).run_test()
 
+    def test_concat_threshold_update(self):
+        """
+        This test checks that the "broken" node (node without output) is being
+        removed from the graph during quantization.
+        """
```
Comment (reuvenperetz) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:588]: Please remove/fix this comment.

File: tests/pytorch_tests/model_tests/feature_models/concat_threshold_test.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# Copyright 2022 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (reuvenperetz) [inline tests/pytorch_tests/model_tests/feature_models/concat_threshold_test.py:1]: Please fix license year

File: tests/keras_tests/feature_networks_tests/feature_networks/concatination_threshold_update.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2021 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (reuvenperetz) [inline tests/keras_tests/feature_networks_tests/feature_networks/concatination_threshold_update.py:1]: Please fix license year

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py (hunk 4 above)
Comment (samuel-wj-chapman) [reply model_compression_toolkit/core/keras/graph_substitutions/substitutions/concat_threshold_update.py]: changed code to leave the case of MP. will add a fix if results for the method are good for ptq.

File: model_compression_toolkit/core/common/quantization/quantization_config.py (hunk 6 above)
Comment (samuel-wj-chapman) [reply model_compression_toolkit/core/common/quantization/quantization_config.py:66]: docs added

**Comment by samuel-wj-chapman** [comment]
made the reinvent changes. let me know if i have filled in all the correct documentation.

**Comment by samuel-wj-chapman** [comment]
closing pr due to commit issues re opoened as pr 1036
